New: Add id-length rule (fixes #2784)

BYK opened the pull request:
Fixes #2784

BYK commented on tests/lib/rules/id-length.js:
```diff
@@ -0,0 +1,66 @@
+        { code: "var obj = {a: 1, bc: 2};", errors: [{ message: "Identifier name 'a' is too short. (< 2)", type: "Identifier"}] },
+        { code: "try { blah(); } catch (e) { /* pass */ }", errors: [{ message: "Identifier name 'e' is too short. (< 2)", type: "Identifier"}] },
+        { code: "var handler = function (e) {};", errors: [{ message: "Identifier name 'e' is too short. (< 2)", type: "Identifier"}] },
+        { code: "var handler = function ($e) {};", errors: [{ message: "Identifier name '$e' is too short. (< 2)", type: "Identifier"}] },
```
These type of errors might be a bit confusing since it says `$e` is too short (< 2), where it actually is two characters and we check the "bare" length after stripping prefixes and suffixes. I couldn't find a succinct enough error message to express this subtlety :(

BYK commented on conf/eslint.json:
```diff
@@ -137,6 +137,7 @@
         "guard-for-in": 0,
         "handle-callback-err": 0,
         "indent": 0,
+        "id-length": 0,
```
`id-len` might be an even better name.

gyandeeps commented on lib/rules/id-length.js:
```diff
@@ -0,0 +1,108 @@
+                        "ObjectExpression",
+                        "FunctionExpression",
+                        "FunctionDeclaration",
+                        "CatchClause"
```
Please mention this in docs also that these are the available options for `disableFor`. Makes it easier for people.

gyandeeps commented:
Please format your commit message as per our guidelines: http://eslint.org/docs/developer-guide/contributing

BYK commented:
@gyandeeps done! :)

nzakas commented on docs/rules/id-length.md:
````diff
@@ -0,0 +1,77 @@
+
+```js
+// id-length: 1
+var x = 5;
````
Can you add some blank lines in between each example here?

nzakas commented on docs/rules/id-length.md:
````diff
@@ -0,0 +1,77 @@
+```js
+// id-length: 1
+var num = 5;
+function _func() { return 42; }
````
Same here

nzakas commented on docs/rules/id-length.md:
```diff
@@ -0,0 +1,77 @@
+* **min** *(default: 2)*: The minimum number of characters an identifier name should be, after it is stripped from it is prefixes and suffixes
+* **max** *(default: Infinity)*: The maximum number of characters an identifier name should be, after it is stripped from it is prefixes and suffixes
+* **exceptions** *(default: ["$", "_", "i", "j"])*: An array of identifier names that the rule should not apply to
+* **disableFor** *(default: [])*: An array of node types that the rule should not be applied to. Available types are:
```
Is this necessary? I'm generally not a big fan of requiring node types in configuration and I don't see a clear use case for this.

BYK replied:
I added this for the question you've raised:

>First, I was under the impression that this would be just for variable and function names, but it looks like it's also for properties. I'm not sure it makes sense to bundle those together, and if it does, we need a way to separate them so people can have different rules for variables/functions as compared to properties.

I personally don't use it but looks like we may wanna give people some "relaxation" options. May not be necessary for the first pass as you pointed out.

nzakas commented on docs/rules/id-length.md:
```diff
@@ -0,0 +1,77 @@
+
+This rule is aimed at increasing code readability and maintainability by enforcing an identifier length convention. It will warn on any type of identifier which doesn't conform to length limits (upper and lower). The rule ignores common prefixes and suffixes up to two chars of `_` and `$` in any combination.
+
+It has exceptions for commonly used single-char identifier names such as `i` or `$`. It also allows the programmers to silently by-pass this check by using "quoted" property names or calculated property access to allow potential server-side data requirements.
```
We cannot prepopulate identifiers in this way as there would be no way to say that these should not be allowed. There need to be no exceptions by default.

BYK replied:
I almost agree but not including `i` as a default sounds like a crime. Are you 100% sure about this?

nzakas replied:
Yes

btmills replied:
@nzakas Would default exceptions be acceptable if specifying an exceptions list overrode the defaults, allowing users to remove the default exceptions? A couple examples:

```js
{
    "id-length": 2, // Exceptions default to ["i", "j", "_", "$"]
    "id-length": [2, {
        "exceptions": [] // No exceptions
    }],
    "id-length": [2, {
        "exceptions": ["$"] // Only allow "$"
    }],
    "id-length": [2, {
        "exceptions": ["i", "j", "_", "$", "x"] // Defaults plus "x"
    }]
}
```

I'm proposing this because I like those defaults, though I completely agree it should be possible to override them. (Conveniently, the modification to support this would only require changing the empty fallback `[]` to include the defaults.)

nzakas commented:
This is a good start. One of the challenging things about core rules is that they need to be consistently generic, and this rule in its current incarnation has a lot of gotchas.

First, I was under the impression that this would be just for variable and function names, but it looks like it's also for properties. I'm not sure it makes sense to bundle those together, and if it does, we need a way to separate them so people can have different rules for variables/functions as compared to properties.

Next, we can't have any special treatment for anything baked in. That means no predefined exceptions and no special treatment for prefixes. The length must always be treated as absolute.

Keep in mind, it's easy to add things we miss later, but it's nearly impossible to remove things once they're introduced, so we always start with very simple rules and add options as we find necessary vs. trying to figure out every use case ahead of time and ship a big thing.

So, I think we should strip this rule down to the minimum implementation as a first pass and wait for feedback before adding in all the bells and whistles.

michaelficarra commented:
Agreed and well put, @nzakas.

BYK commented:
>First, I was under the impression that this would be just for variable and function names, but it looks like it's also for properties. I'm not sure it makes sense to bundle those together, and if it does, we need a way to separate them so people can have different rules for variables/functions as compared to properties.

I included properties because `exports` pattern allows people to assign any "variables" or "functions" to objects like this and they are essentially properties. That's why this rule became `id-length` instead of `var-length`. Not sure if we really need a way to define separate rules for these but it sounds double through a map: `"ExpressionType": { /* expression option overrides */ }`. Sounds a bit more complicated thoudh.

>Next, we can't have any special treatment for anything baked in. That means no predefined exceptions and no special treatment for prefixes. The length must always be treated as absolute.

I was thinking the same but when I started implementing this rule, I looked at other rules and took this idea from here: https://github.com/eslint/eslint/blob/master/lib/rules/camelcase.js#L53

>Keep in mind, it's easy to add things we miss later, but it's nearly impossible to remove things once they're introduced, so we always start with very simple rules and add options as we find necessary vs. trying to figure out every use case ahead of time and ship a big thing.

>So, I think we should strip this rule down to the minimum implementation as a first pass and wait for feedback before adding in all the bells and whistles.

Totally agree but I also think things should come with sensible defaults. Of course there are things we can strip but I want you to reconsider them under the light of my reasoning. Then I'll follow the path you suggest.

Sounds like a deal?

nzakas commented:
I understand your points, however, my original comments stand as my opinion for moving forward. Sensible defaults are not universal (as we have learned the hard way over time), and once introduced, are hard to change. The best option is always to start with no default exceptions and wait for feedback to tell us what is or isn't painful enough to make a change.

BYK commented:
>The best option is always to start with no default exceptions and wait for feedback to tell us what is or isn't painful enough to make a change.

Alright then. Changes are upcoming.

BYK commented:
Updated:
* Make examples readable
* Remove default exceptions
* Remove prefix/suffix stripping
* Remove `disableFor`

nzakas commented on docs/rules/id-length.md:
````diff
@@ -0,0 +1,84 @@
+Identifier names such as `e`, `x`, `_t` or sometimes very long ones like `hashGeneratorResultOutputContainerObject` which usually makes the code harder to read and less maintainable. To prevent this, some code bases may enforce a minimum and/or maximum identifier length. (usually min 2-chars)
+
+```js
+// id-length: 1  // minimum 2-chars
````
this should be "min", correct?

nzakas replied:
Ooh I see, I think this is confusing as-is. I dying think people will understand that 1 is not a length. I'd specify the options throughout.

nzakas commented on tests/lib/rules/id-length.js:
```diff
@@ -0,0 +1,57 @@
+        "var _a = 2",
+        "var _ad$$ = new $;",
+        "var xyz = new ΣΣ();",
+        { code: "var x = Foo(42)", args: [1, {"min": 1}] },
```
Can you use `options` instead of args? We are deprecating args. `options` is the same except you don't need the first number (that's not actually used anywhere)

nzakas commented:
This looks good, just a couple notes.

What do people think about id-length vs. id-len?

ilyavolodin commented:
I think `id-length` is short enough.

BYK commented:
Updated and rebased. Sorry it took a while, life intervened :)

Not sure if you'd like the final version of docs though. If you can be more specific about how to improve it, it'd really help.

ilyavolodin commented on lib/rules/id-length.js:
```diff
@@ -0,0 +1,87 @@
+/**
+ * @fileoverview Rule that warns when identifier names are shorter or longer
+ *               than the values provided in configuration.
+ * @author Burak Yigit Kaya aka BYK
```
Missing copyright

BYK replied:
I don't want a copyright. There are rules (some by you) missing copyright :)

ilyavolodin replied:
Those rules by me were created way before we started adding copyrights:-) But if you don't want to add one, that's perfectly fine.

BYK replied:
I mean I want anyone to do whatever he or she wants with this code and a copyright implies that they need to consult to me or something right? That's why I'm resisting. If my understanding is not correct though, I'd add it especially for consistency :)

xlong328 replied:
xlong328@qq.com

platinumazure replied:
Not true that copyright automatically implies restriction of use of code. Copyright just means that this particular code snippet is definitely owned (intellectually) by you. It is possible to use a software license which asserts copyright but allows distribution and modification under reasonable conditions. Check out the project's license for more information.

nzakas replied:
@byk Basically, you own own whatever code you write. The CLA says you are licensing that code to ESLint so we can include it under our license terms. It's not restricting anyone's ability to use the code, it's just asserting that you created it AND that's what gives you the right to allow us to use it.

BYK replied:
Alright, adding it.

ilyavolodin commented on lib/rules/id-length.js:
```diff
@@ -0,0 +1,87 @@
+
+module.exports = function (context) {
+    var options = context.options[0] || {};
+    var minLength = typeof options.min === "number" ? options.min : 2;
```
Don't need to check for number, schema is going to blow up if non-number is going to be passed in.

ilyavolodin commented on lib/rules/id-length.js:
```diff
@@ -0,0 +1,87 @@
+    var options = context.options[0] || {};
+    var minLength = typeof options.min === "number" ? options.min : 2;
+    var maxLength = typeof options.max === "number" ? options.max : Infinity;
+    var exceptions = (Array.isArray(options.exceptions) ? options.exceptions : [])
```
Same with `Array.isArray`

ilyavolodin commented on docs/rules/id-length.md:
```diff
@@ -0,0 +1,84 @@
+# Limit minimum and maximum length for identifiers (id-length)
+
+Identifier names such as `e`, `x`, `_t` or sometimes very long ones like `hashGeneratorResultOutputContainerObject` which usually makes the code harder to read and less maintainable. To prevent this, some code bases may enforce a minimum and/or maximum identifier length. (usually min 2-chars)
```
I would restructure this sentence. It just doesn't flow.

BYK commented:
@ilyavolodin - updated!

ilyavolodin commented:
There are some linting errors in the code. That's why builds are failing:
```
lib/rules/id-length.js
  15:37  error  Unexpected use of undefined  no-undefined
  16:37  error  Unexpected use of undefined  no-undefined
```

BYK commented:
>There are some linting errors in the code. That's why builds are failing:

Oh boy, I updated this without running `npm test` lol. Sorry about that.

BYK commented:
@ilyavolodin  - I did fix those `no-undefined` violations, rebased and now seeing many many lint failures. I'm guessing this is due to the recently landed changes around default rules.

Should I wait until they are fixed on master and rebase or is this okay to merge now?

mysticatea commented on lib/rules/id-length.js:
```diff
@@ -0,0 +1,87 @@
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = function (context) {
```
I guess new rule config for ESLint itself disallows spaces before the `(` of the argument list.

ilyavolodin commented:
@BYK Master is error free currently. All of the 1.0.0 patches has already landed into master, so it looks like those errors are valid. It looks like all of them are because of the slight difference in coding style. We do not use spaces before function parents and it looks like you do.

BYK commented:
@ilyavolodin - interesting, these didn't happen before. Also I was getting many errors on other files even including the `Makefile.js`. Let me try again.

mysticatea commented:
Ah, maybe, it's related with #3005 
On Windows, `npm test` reports many errors by `spaced-comments`

BYK commented:
@mysticatea oh yeah that's it. Thanks for the pointer!

ilyavolodin commented:
I think this is good to merge after rebase.

BYK commented:
Rebased again :)

ilyavolodin commented:
@BYK Apparently code coverage for branches felt below the threshold.

nzakas commented on tests/lib/rules/id-length.js:
```diff
@@ -0,0 +1,57 @@
+//------------------------------------------------------------------------------
+
+var eslint = require("../../../lib/eslint"),
+    ESLintTester = require("eslint-tester");
```
Can you switch this to be:

```js
ESLintTester = require('../../../lib/testers/eslint-tester')
```

BYK replied:
Done.

BYK commented:
Alright, rebased, made requested changes + improved the branch coverage.

IanVS commented:
Thanks for your persistence, @BYK

BYK commented:
>Thanks for your persistence, @BYK

Not gonna give up ;)

nzakas commented:
Lgtm
